File: python/145.Binary Tree Postorder Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        tmpNode = root
        preVisitNode = None
        rnt = []

        while(tmpNode!= None or len(stack)!=0):

            while(tmpNode != None):
                stack.append(tmpNode)
                tmpNode = tmpNode.left

            if len(stack)!=0:
                tmpNode = stack[len(stack)-1]
                if tmpNode.right == None or tmpNode.right == preVisitNode:
                    stack.pop()
                    rnt.append(tmpNode.val)
                    preVisitNode = tmpNode
                    tmpNode = None
                else:
                    tmpNode = tmpNode.right   
        return rnt
        

if __name__ == '__main__':
    root = TreeNode(1)
    root.right = TreeNode(2)
    root.right.left = TreeNode(3)
    print(Solution().postorderTraversal(root))

File: cpp/145.Binary Tree Postorder Traversal.cpp
//Definition for a binary tree node.
# include <stack>
# include <vector>
# include <stdlib.h>
# include <iostream>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
    TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        TreeNode *tmpNode = root;
        vector<int> ret;
        if (tmpNode == NULL)
        	return ret;
		TreeNode *preVisitNode = NULL;
        stack<TreeNode*> S;
        while(tmpNode!=NULL || !S.empty()){
        	while(tmpNode!=NULL){
        		S.push(tmpNode);
        		tmpNode = tmpNode->left;
			}
			if (!S.empty()){
				tmpNode = S.top();
				if (tmpNode->right==NULL || preVisitNode==tmpNode->right){
					ret.push_back(tmpNode->val);
					preVisitNode = tmpNode;
					tmpNode = NULL;
					S.pop();
				}
				else
					tmpNode = tmpNode->right;
			}
		}
		return ret;
    }
};

int main(){
	TreeNode *root =  new TreeNode(1);
	root->right =  new TreeNode(2);
	root->right->left = new TreeNode(3);
	Solution s = Solution();
	s.postorderTraversal(root);
}


